Drillsqueak Beta 3.0.0b3

theunkn0wn1 opened the pull request:
this is a continuation of #165 , fixing several bugs, added lots of tests, other backend magic etc.

there is honestly quite a bit to this pr... :sweat_smile: 

Note this is what is currently deployed to drillsqueak.

theunkn0wn1 committed:
Squash initial commit of API integration

theunkn0wn1 committed:
Squash add tests from old branch

theunkn0wn1 committed:
any 2xx is a success.

theunkn0wn1 committed:
Add create rescue test
 - fix some broken stuff in plain_rescue_fx

theunkn0wn1 committed:
WIP unidentified rats

theunkn0wn1 committed:
`ApiABC` -> `FuelratsApiABC`

theunkn0wn1 committed:
Began tying the board into the API protocol

theunkn0wn1 committed:
pep8

theunkn0wn1 committed:
Technically all the API tests are actually integration tests.

theunkn0wn1 committed:
Simplify modify_rescue

theunkn0wn1 committed:
Refactor/simplified MFD using attr.s
 - removed a bunch of now strictly invalid tests

theunkn0wn1 committed:
Added Board integration tests

theunkn0wn1 committed:
Add data attribute converter

theunkn0wn1 committed:
Fix check condition for `Rescue.mark_delete`

theunkn0wn1 committed:
Parse `attribute.data` from API

theunkn0wn1 committed:
Simplify add_rat boolean.

theunkn0wn1 committed:
Black'ed new code

theunkn0wn1 committed:
Too many blank lines

theunkn0wn1 committed:
Went off the deep end of converters.

theunkn0wn1 committed:
Culled unused artefacts

theunkn0wn1 committed:
`RatBoard.create_rescue` is now not a context manager.

theunkn0wn1 committed:
`RatBoard.create_rescue` is now not a context manager. pt2.

theunkn0wn1 committed:
renamed `Data` to `InternalData` to be more representative

theunkn0wn1 committed:
Type annotate `Board` constructor

theunkn0wn1 committed:
Don't overwrite Rescue.mark_for_deletion if the new MFD is invalid

theunkn0wn1 committed:
Simplify MFD validity check

theunkn0wn1 committed:
removed superfluous  check, differing to `attr.s` validator

shatteredbeam committed:
Initial

shatteredbeam committed:
pre-breakpoint issue.

shatteredbeam committed:
active, Assign, clear

theunkn0wn1 committed:
store clients casefolded, lookup is now caseless.
 - allow `Rescue` objects into `modify_rescue`

shatteredbeam committed:
refactor validation into its own helper method, added more commands.

theunkn0wn1 committed:
Protect board data from concurrent IO

theunkn0wn1 committed:
remove rescue needs to be protected, too

shatteredbeam committed:
Add more functionality to grab, enable injected case creation.

theunkn0wn1 committed:
[SPARK-159] Redo Quotation with more modern tech. Fixed SPARK-159 in the process.

 - invalidated a bunch of tests too

theunkn0wn1 committed:
[SPARK-159] Added missing documentation

shatteredbeam committed:
More work on commands.

shatteredbeam committed:
More commands, the sequel

shatteredbeam committed:
Added start time property to mechaclient.py, for later user in uptime command.

theunkn0wn1 committed:
Rescue shouldn't care about the source of a Rat object
 - Rescue's (add/remove)_rat now simply accept a Rat object
 - no fancy cache lookups
 - removed outdated tests that verified the old behavior
 - added property to rat to tell if its an identified object

theunkn0wn1 committed:
Fix over-indent

shatteredbeam committed:
More commands, electric bugaloo

shatteredbeam committed:
First draft of commands.  Rescue management commands/dispatch commands completed.

shatteredbeam committed:
First Draft for PR

theunkn0wn1 committed:
fix missing return in bad sys usage

theunkn0wn1 committed:
fix underlying bug in board implementation

theunkn0wn1 committed:
added new command implemenations to __all_

theunkn0wn1 committed:
!mfd should remove rescue from active tracking

theunkn0wn1 committed:
Fix !clear not calling the proper remove method

theunkn0wn1 committed:
regex strings should always be escaped

theunkn0wn1 committed:
use relative imports to make later refactors easier

theunkn0wn1 committed:
use relative imports to make later refactors easier

theunkn0wn1 committed:
use relative imports to make later refactors easier

theunkn0wn1 committed:
create rescue isn't a context manager any more

theunkn0wn1 committed:
unnecessary string concat

theunkn0wn1 committed:
fix unused variable and made !unassign consistent with production

theunkn0wn1 committed:
implemented !list

theunkn0wn1 committed:
fixed formatting error in ratlib

theunkn0wn1 committed:
actually fixed formatting

theunkn0wn1 committed:
actually fixed formatting

theunkn0wn1 committed:
Fixed rendering of unidentified rats assigned to a case

theunkn0wn1 committed:
failsafe check that the client is actually still tracked

theunkn0wn1 committed:
fixed some edge cases in !grab

theunkn0wn1 committed:
began binding starsystems

theunkn0wn1 committed:
attempted to implement !landmark-near

theunkn0wn1 committed:
case insensitive lookup

theunkn0wn1 committed:
use light blue instead of blue as per request

theunkn0wn1 committed:
fixed !xb not actually assigning XB

theunkn0wn1 committed:
additional starsystem diagnostics

theunkn0wn1 committed:
fixed null dereference in rsig parser.

theunkn0wn1 committed:
fixed missing rsig hooks

theunkn0wn1 committed:
bypassed rule system.

theunkn0wn1 committed:
fixed !ircnick index error

theunkn0wn1 committed:
fixed md word count check

theunkn0wn1 committed:
Some fixes for `landmark near`

theunkn0wn1 committed:
not allowed to clear to an unidentified rat, not allowed to clear to an unassigned rat.

theunkn0wn1 committed:
added on_invite hook

theunkn0wn1 committed:
began implementing command unit tests, fixed duplicate name
 - why wasn't this caught by the linter?

theunkn0wn1 committed:
fixed order-of-operations error in rescue causing `active=False` in the constructor to do nothing

theunkn0wn1 committed:
also test unassign behavior.

theunkn0wn1 committed:
Added test for clear

theunkn0wn1 committed:
More tests for clear

theunkn0wn1 committed:
test grab create

theunkn0wn1 committed:
Deduplication x2 in !(pc/xb/ps)

theunkn0wn1 committed:
fixed case sensitivity in !assign

theunkn0wn1 committed:
snickers and cake

theunkn0wn1 committed:
fix some case sensitivities

theunkn0wn1 committed:
implement version command

theunkn0wn1 committed:
Simplified `Rat` class greatly

theunkn0wn1 committed:
Defused time bomb.

theunkn0wn1 committed:
Unionized !list and !quotes output format so code could be reused

theunkn0wn1 committed:
started doing some fuzz testing

theunkn0wn1 committed:
adjusted inject fuzz test to use new input restrictions
 - to make the test data more sane relative to expected prudction inputs

theunkn0wn1 committed:
simplified fuzzing assumptions as the filters are applied upstream

theunkn0wn1 committed:
Added check for CR warnings

theunkn0wn1 committed:
use irc nickname instead of client name

theunkn0wn1 committed:
Added strategy for generating rescues

theunkn0wn1 committed:
Additional !sub tests and resulting bug fixes.

theunkn0wn1 committed:
Added a repr to rescue to facilitate reading hypothesis test cases

theunkn0wn1 committed:
ensure all appended rescues have a board index when being appended
 - this field cannot be null going into the board...
 - thanks hypothesis!

theunkn0wn1 committed:
Tie in a test fact (temporary)

theunkn0wn1 committed:
actually load the administration module...

theunkn0wn1 committed:
Sneak a get-fact into debug. expanded fact manager logging

theunkn0wn1 committed:
Bad counter! bad!

theunkn0wn1 committed:
add fact handling to front end

theunkn0wn1 committed:
missing await :facepalm:

theunkn0wn1 committed:
additional logging

theunkn0wn1 committed:
More logging...

theunkn0wn1 committed:
the command invocation is the thing itself

theunkn0wn1 committed:
Sanity check around fact handling

theunkn0wn1 committed:
English mother ratter, do you speak it!

Eearslya commented on src/commands/_list_flags.py:
```diff
@@ -0,0 +1,43 @@
+        """
+        Show assigned rats per rescue
+        """
+        filter_unassigned_rescues = "u" in argument
```
[Style, Suggestion] Newlines between docstrings to make it more obvious which string is referencing what.

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,700 @@
+from ..packages.rescue import Rescue
+from ..packages.utils import Platforms, Status
+
+_TIME_RE = re.compile(r'(\d+)[: ](\d+)')
```
[Style] Might deserve a docstring describing what this is for.

theunkn0wn1 replied:
I agree. @shatteredbeam what *does* this do?

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,700 @@
+
+
+# User input validation helper
+def _validate(ctx: Context, validate: str) -> typing.Optional[Rescue]:
```
[Style] Missing docstring.

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,700 @@
+    rescue = _validate(ctx, ctx.words[1])
+
+    # Only set First Limpet if it was specified.
+    if len(ctx.words) == 3:
```
[Question] Maybe we should throw a usage error if the command has more than 3 words, too? Just to be safe. e.g. `!close 3 unknown [pc]` would close the case with no FL.

theunkn0wn1 replied:
yeah, this one would be better to throw a usage error; you can't close a case to more than one FL. After all, its not called a _first_ limpet for nothing.

theunkn0wn1 committed:
Catch any database error and squash it.

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,699 @@
+        return await ctx.reply(f"{case.client}'s case opened with {last_message!r}")
+
+    if ctx.words[1].casefold() in ctx.bot.last_user_message:
+        last_message = ctx.bot.last_user_message[ctx.words[1].casefold()]
```
[Error?] I'm trying to look at this as best as I can, but this seems to be redundant because of L274?

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,699 @@
+        last_message = ctx.bot.last_user_message[ctx.words[1].casefold()]
+    elif int(ctx.words[1]) in ctx.bot.board:
+        if rescue.client.casefold() in ctx.bot.last_user_message:
+            last_message = last_message
```
[Error] What exactly does this do...?

theunkn0wn1 replied:
good question. looking at the delta i introduced that i can't make sense of why i changed it either. reverting..

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,699 @@
+            for keyword in ctx.words_eol[2].split():
+                if keyword.upper() in {item.value for item in Platforms}:
+                    case.platform = Platforms[keyword.upper()]
+                if keyword.casefold() == "cr" or "code red" in keyword.casefold() \
```
[Error] If the `words_eol` is split on whitespace, `"code red"` is not a possible result.

theunkn0wn1 replied:
that isnt right *writes up counter example* oh, so it is.

Fixed.

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,699 @@
+        await ctx.reply(f"Set IRC Nickname to {case.irc_nickname!r}")
+
+    if new_name.casefold() != rescue.client.casefold():
+        await ctx.reply("Caution: IRC Nickname does not match CMDR Name.")
```
[Question] This warning seems redundant. Isn't the whole point of this command to set IRC/CMDR names to different values?

theunkn0wn1 replied:
no. generally the IRC nickname matches the Commander name, and it is thus a warning in mecha2 if these two values diverge.

Eearslya replied:
That seems like a fair warning when the case is first announced, but the entire point of this command is to change the IRC Nickname away from the CMDR name, is it not? Since the client usually joins with their CMDR Name as their IRC Name to begin with (barring any forbidden characters).

Eearslya commented on src/commands/deletion_management.py:
```diff
@@ -0,0 +1,85 @@
+
+
+# User input validation helper
+def _validate(ctx: Context, validate: str) -> Optional[Rescue]:
```
[Error] Duplicate from `case_management.py`, consider moving it?

shatteredbeam replied:
Interesting, in my version from awhile back this move was already done.

Eearslya commented on src/mechaclient.py:
```diff
@@ -70,6 +75,9 @@ def __init__(self, *args, mecha_config=None, **kwargs):
     #
     # def on_join(self, channel, user):
     #     super().on_join(channel, user)
+    async def on_invite(self, channel, by):
```
[Suggestion] Consider limiting this by inviter VHost, if possible?

theunkn0wn1 replied:
yeah... this should probably be behind permission checks.

theunkn0wn1 replied:
marking as  SPARK-163, will fix

Eearslya commented on src/packages/board/board.py:
```diff
@@ -155,85 +193,148 @@ def free_case_number(self) -> int:
+        Args:
+            rescue (Rescue): object to append
+            overwrite(bool): overwrite existing cases
+    """
```
[Style] Docstring indentation

theunkn0wn1 committed:
Black -l 102 src/commands

theunkn0wn1 committed:
black -l 102 src/commands

Eearslya commented on tests/integration/api/test_mock_v3.py:
```diff
@@ -0,0 +1,29 @@
+    target = UUID("c1ff8456-5c51-47b1-8d0e-be6104840820")
+    rescue_sop_fx._api_id = target
+
+    await mock_fuelrats_api_fx.update_rescue(rescue_sop_fx)
```
[Error] No assertions? If this test is meant to just assert that no exceptions are raised, it should probably say so explicitly.

theunkn0wn1 replied:
i... honestly don't even remember. I also don't know how much of this implementation is even being preserved. Looking at the implementation, this was for the mockup v3 rather than the actual api v3. i may just gut these tests entirely

Eearslya commented on tests/integration/test_board.py:
```diff
@@ -0,0 +1,21 @@
+    rescue = await  board_online_fx.create_rescue(
+        "UNIT_TEST", "ki", "UNIT_TEST", status=Status.OPEN, platform=Platforms.PC
+    )
+    ...
```
[Style] Unnecessary pass

Eearslya reviewed (commented):
Some changes I'd like to see and questions answered, but nothing explicitly preventing merge imo.

theunkn0wn1 committed:
have ratmamma use the generated case's data instead of its creation data
 - more accurate this way

theunkn0wn1 committed:
Temporary resolution to the r@sig thing
 - more permanent solution incoming...

theunkn0wn1 committed:
added manual check because humans arn't consistent

shatteredbeam commented on src/mechaclient.py:
```diff
@@ -126,7 +135,7 @@ def fact_manager(self) -> FactManager:
         """
         if not self._fact_manager:
-            self._fact_manager = FactManager()  # Instantiate Global Fact Manager
+            self._fact_manager = FactManager(fact_table="converted_facts")  # Instantiate Global Fact Manager
```
[Error] this as discussed is kind of hacky.  The fact manager wasn't designed to be given a table for security reasons, but rather than hard code the name it needs to be moved to a config option.

shatteredbeam replied:
I also agreed to take care of this, so this should not be considered blocking.

theunkn0wn1 replied:
SPARK-167

shatteredbeam commented on src/packages/commands/rat_command.py:
```diff
@@ -76,10 +81,46 @@ class NameCollisionException(CommandException):
+        fact = await context.bot.fact_manager.find(fact, lang)
+        await context.reply(fact.message)
+        return True
+    except psycopg2.Error:
```
Can we use a more specific error handler here? I feel this is too broad.

theunkn0wn1 replied:
Nope! not without having some way of asking the FM if it has a connection or not.

shatteredbeam replied:
You can.  if the pool is None or in an error state there's no active connection.  I believe there's also a 'connected' property I added.

shatteredbeam replied:
We discussed this, and decided there was a better way - by adding a method to check if the current connection exists.  This should also be considered non-blocking and out of scope.

theunkn0wn1 replied:
For the sake of keeping the drillsqueak beta alive, we will leave this change in for now. 
As per our discussion, @shatteredbeam will provide better methods for achieving this end without such an ugly exception clause in a later PR.

theunkn0wn1 replied:
SPARK-164

shatteredbeam commented on src/packages/database/database_manager.py:
```diff
@@ -174,6 +174,7 @@ def __init__(self,
             connection.set_client_encoding('utf-8')
             # Create cursor, and execute the query.
             with connection.cursor() as cursor:
+                logger.debug("executing query {}", query)
```
this has security concerns, if you're exposing the entire query.

theunkn0wn1 replied:
placed behind a `__debug__` check.

Eearslya commented on src/packages/ratmama/ratmama_parser.py:
```diff
@@ -13,17 +13,20 @@
 
 
-_config: Dict = {}
+class _RatmammaConfig(TypedDict):
```
[Error?] Are we going with `Ratmama` or `Ratmamma`? The file-name seems to indicate the former.

theunkn0wn1 replied:
... that wasn't even supposed to be comitted yet darnit.
will fix

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,725 @@
+from ..packages.rescue import Rescue
+from ..packages.utils import Platforms, Status
+
+_TIME_RE = re.compile(r"(\d+)[: ](\d+)")
```
```suggestion
_TIME_RE = re.compile(r"(\d+)[: ](\d+)")
"""
Regex matcher used to find a time within a string. Used to determine
if a newly-submitted case is code red or not.
"""
```

shatteredbeam replied:
I was just working on answering this, but yes.  this is just a simple way to see if time was included in an inject, and then to auto-mark the case CR if included.

theunkn0wn1 committed:
partial revert

theunkn0wn1 committed:
remove redundant pass

theunkn0wn1 committed:
don't log queries in production builds

theunkn0wn1 committed:
docstring for time detecting regex

Co-Authored-By: Eearslya Sleiarion <eearslya@gmail.com>

theunkn0wn1 committed:
fuzz testing for close, refactor custom strategies to match hypothesis structure

theunkn0wn1 committed:
extended validation for !close to prevent invalid usages..

theunkn0wn1 committed:
Resolved spelling error

theunkn0wn1 committed:
made PyLint happy. black -l 103 over a couple files + a manual patch

theunkn0wn1 committed:
manually reduced the length of a line black refused to touch

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,732 @@
+
+    if new_name.casefold() != rescue.client.casefold():
+        await ctx.reply(
+            f"Caution: IRC Nickname {new_name!r}does not match CMDR Name {rescue.client.casefold()!r}."
```
[Style] Missing space after `new_name`.

Eearslya commented on src/commands/case_management.py:
```diff
@@ -0,0 +1,732 @@
+    buffer.write(f"{len(rescue_collection)} active cases. ")
+    for rescue in rescue_collection:
+        buffer.write(format(rescue, format_specifiers))
+        buffer.write("\n")
```
[Error?] Looking at Mecha2's current `list` functionality, isn't there only meant to be a newline between cases if additional flags are given? (e.g. show assigned rats)

Eearslya commented on src/commands/starsystems.py:
```diff
@@ -0,0 +1,35 @@
+
+    logger.debug("found system {}, acquiring nearest landmark...", found)
+    nearest_landmark = await ctx.bot.galaxy.find_nearest_landmark(found)
+    return await ctx.reply(f"{nearest_landmark}")
```
[Suggestion] `find_nearest_landmark` also returns the distance as a second return value, suggest we pass this info along.

theunkn0wn1 replied:
[error] actually implement this routine properly.
whats currently here is test code

Eearslya commented on src/packages/rescue/rat_rescue.py:
```diff
@@ -837,3 +784,76 @@ def change(self):
+                if self.platform is Platforms.XB:
+                    buffer.write(color(base, Colors.GREEN))
+                elif self.platform is Platforms.PS:
+                    buffer.write(color(base, Colors.LIGHT_BLUE))
```
[Error] Missing case for PC.

theunkn0wn1 replied:
you know i looked at this a good dozen times, and only now am i seeing that. the Else statement on l824 got be confused...

Eearslya commented on tests/strategies.py:
```diff
@@ -0,0 +1,85 @@
+                                lambda case: case.client))
+
+
+""" A list of rescues """
```
[Style] What is this docstring attached to? If it's the above `rescues` function, it seems odd to have two newlines between them.

theunkn0wn1 replied:
its orphaned. removing.

theunkn0wn1 committed:
fixing missing case for PC

theunkn0wn1 committed:
resolved orphaned docstring

theunkn0wn1 committed:
use drillsignal for now, rules will need reworking to facilitate configuration

theunkn0wn1 committed:
bump version to b4

theunkn0wn1 committed:
lets see if removing the branch specifier helps

theunkn0wn1 committed:
happy now codecov?

shatteredbeam committed:
Fact Manager will now pull the table names for the fact and fact log tables directly from configuration file.  A passed property to the fact manager will override the configuration file.

shatteredbeam committed:
Added 'is_connected' function to test connectivity on demand.  will return false and raise if no connectivity.

Expanded on error handling for detecting database errors, programmatic errors or an integrity violation.

theunkn0wn1 committed:
resolve SPARK-171

theunkn0wn1 committed:
bump pyproject.toml version

theunkn0wn1 committed:
you weren't supposed to be commited yet!

theunkn0wn1 committed:
No, seriously, exit stage left now. Your part comes later ;)

Eearslya commented on src/mechaclient.py:
```diff
@@ -102,6 +102,10 @@ def __init__(self, *args, mecha_config=None, **kwargs):
             self._last_user_message[user.casefold()] = sanitized_message  # Store sanitized message
             ctx = await Context.from_message(self, channel, user, sanitized_message)
+            if not ctx.words:
+                logger.trace("ignoring empty message")
```
[Suggestion] As this came up as a bug, should we add a regression test?

theunkn0wn1 replied:
agreed.

theunkn0wn1 committed:
Merge remote-tracking branch 'shattered/unk/PR175' into core/commands

theunkn0wn1 committed:
fixed impossible constraint

theunkn0wn1 committed:
deprecate the API fixture for now

theunkn0wn1 committed:
Merge remote-tracking branch 'origin/core/commands' into core/commands

theunkn0wn1 committed:
take a stab at SPARK-172

theunkn0wn1 committed:
Version bump

theunkn0wn1 committed:
Black -l 103 src/packages/database/database_manager.py

theunkn0wn1 committed:
actually resolve the lint error.
was in a comment so Black refused to touch it

theunkn0wn1 committed:
Also the lang extensions, please.

theunkn0wn1 committed:
version bump

theunkn0wn1 committed:
reply to mentioned users in fact invocations

theunkn0wn1 committed:
version bump

theunkn0wn1 committed:
SPARK-163 place on_invite behind a permissions check.

theunkn0wn1 committed:
Black -l 102 src/mechaclient.py

theunkn0wn1 committed:
Bump version

theunkn0wn1 committed:
diagnostic

theunkn0wn1 committed:
Fixed invite hook failing at its one job.

theunkn0wn1 committed:
version bump

shatteredbeam reviewed (approved):
👍

Eearslya reviewed (approved):
Good enough to be getting on with. Polishing can be done later.
